pkg/maps/ctmap/ctmap.go

```diff
@@ -831,3 +875,45 @@ func calculateInterval(prevInterval time.Duration, maxDeleteRatio float64) (inte
 
 	return
 }
+
+func CalculateCTMapPressure(ctx context.Context, allMaps []*Map) error {
+	controller.NewManager().UpdateController("ct-map-pressure", controller.ControllerParams{
+		Group: controller.Group{
+			Name: "ct-map-pressure",
+		},
+		DoFunc: func(context.Context) error {
+			for _, m := range allMaps {
+				path, err := OpenCTMap(m)
+				if err != nil {
+					msg := "Skipping CT map pressure calculation"
+					scopedLog := log.WithError(err).WithField(logfields.Path, path)
+					if os.IsNotExist(err) {
+						scopedLog.Debug(msg)
+					} else {
+						scopedLog.Warn(msg)
+					}
+					continue
+				}
+				defer m.Close()
+
+				l, err := m.DumpAll()
+				switch {
+				case l > 0 && errors.Is(err, ebpf.ErrKeyNotExist):
+					// Kernel returns "not exist" when the batch lookup completes.
+				case err != nil:
+					return err
+				}
+				log.Infof("chris debug: dumped keys %d of CT map", l)
+
+				m.UpdatePressureMetricWithSize(int32(l))
+			}
+			return nil
+		},
+		// TODO: Figure out what the right interval.
+		RunInterval: 30 * time.Second,
+		// TODO: Figure out how to plumb context through or at least how to
+		// stop this controller from hive.
+		Context: ctx,
```

Comment:
I'm having trouble figuring out the hive semantics to plumb through either the `Daemon` context to the ctmap GC cell or to set up the stop hooks. Help appreciated, cc @cilium/sig-foundations.

Reply:
The `pkg/hive/job` package was added for exactly these type of situations. It doesn't yet have the visibility via `ciliumctl status` though.

Here's how this would look like with the `job` package:
```
  func New(..., lc hive.Lifecycle, jobs job.Registry) *GC {
    gc := &GC{...}
    // ...
    // Create a new group of jobs that are started and stopped together.
    group := jobs.NewGroup()

    // Add a timer job that calls calculatePressureMetrics every 30 seconds.
    group.Add(job.Timer(
      "ct-map-pressure", 
       gc.calculatePressureMetrics, // func(context.Context) error
       30 * time.Second, /* TODO: or rather cfg.CTMapPressureInterval? */
    ))

    // Append the job group to the lifecycle.
    lc.Append(group) 

   return gc
  }
```

The job library takes care of the context and cancels it when the application is stopped.

If you'd like to do this with pkg/controller, then you'd need to manage the "module's" context yourself,
e.g. by having for example `func (*GC) Start(hive.HookContext) error` that sets `gc.ctx` and `gc.cancel` and then starts the controllers, and then `func (*GC) Stop(hive.HookContext) error` that calls `gc.cancel` and `RemoveAllAndWait`. In `New() *GC` you'd do `lifecycle.Append(gc)` to add the hooks.

The reason why hive doesn't have an application-wide context is that it's not enough by itself. We also need to wait for things to stop and cleanup (important for tests for example) and further it's often important to keep order when stopping things.

Reply:
Thanks for the context. I've opted for `pkg/controller`, but by using `cancel()` and `RemoveAllAndWait()` to stop and tear down. Credit to @pippolo84 for helping me out last wek.